### ISSUE TITLE
plugin.rb: Option cookbook must be a kind of [String]!  You passed :"jenkins-server"

### ISSUE THREAD

**Issue by nickbroon** [pietervogelaar/chef-cookbook-jenkins-server#1]
I'm seeing a error. Looks to be type mismatch.
My setup is very simple,  a recipe with only:

include_recipe 'jenkins-server::default'

and no node attributes as yet set.

``` ruby
==> default: Recipe Compile Error in /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/stashykins/recipes/default.rb
==> default:
==> default: ================================================================================
==> default:
==> default:
==> default:
==> default:
==> default: Chef::Exceptions::ValidationFailed
==> default:
==> default: ----------------------------------
==> default:
==> default: Option cookbook must be a kind of [String]!  You passed :"jenkins-server".
==> default:
==> default:
==> default:
==> default:   Cookbook Trace:
==> default:   ---------------
==> default:     /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/jenkins-server/recipes/plugins.rb:34:in `block (2 levels) in from_file'
==> default:     /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/jenkins-server/recipes/plugins.rb:33:in `block in from_file'
==> default:     /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/jenkins-server/recipes/plugins.rb:7:in `each'
==> default:     /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/jenkins-server/recipes/plugins.rb:7:in `from_file'
==> default:     /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/jenkins-server/recipes/default.rb:19:in `from_file'
==> default:     /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/stashykins/recipes/default.rb:10:in `from_file'
==> default:
==> default:   Relevant File Content:
==> default:   ----------------------
==> default:   /tmp/vagrant-chef/6e041c8bd39ee7074737ecc884d8b309/cookbooks/jenkins-server/recipes/plugins.rb:
==> default:
==> default:    27:          template_path = options['template_path']
==> default:    28:        else
==> default:    29:          template_path = "#{plugin}.xml"
==> default:    30:        end
==> default:    31:
==> default:    32:        if has_template?(template_source, template_cookbook)
==> default:    33:          template "#{node['jenkins']['master']['home']}/#{template_path}" do
==> default:    34>>           cookbook template_cookbook
==> default:    35:            source template_source
==> default:    36:            mode '0644'
==> default:    37:            owner node['jenkins']['master']['user']
==> default:    38:            group node['jenkins']['master']['group']
==> default:    39:            action :create
==> default:    40:          end
==> default:    41:        else
==> default:    42:          Chef::Log.debug "No template found for source \"#{template_source}\" in cookbook \"#{template_cookbook}\""
==> default:    43:        end
==> default:
==> default:
==> default:   Running handlers:
==> default: [2015-12-11T01:08:33+00:00] ERROR: Running exception handlers
==> default:   Running handlers complete
==> default: [2015-12-11T01:08:33+00:00] ERROR: Exception handlers complete
==> default:   Chef Client failed. 1 resources updated in 05 seconds
==> default: [2015-12-11T01:08:33+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
==> default: [2015-12-11T01:08:33+00:00] ERROR: Option cookbook must be a kind of [String]!  You passed :"jenkins-server".
==> default: [2015-12-11T01:08:33+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
Chef never successfully completed! Any errors should be visible in the
output above. Please fix your recipes so that they properly complete.
```

**Comment by pietervogelaar** [comment]
Can you describe a bit more about your environment? What chef-client version do you use?

**Comment by nickbroon** [comment]
I'm using chef_solo with vagrant, and it's installing Chef Client version 12.5.1 into the debian 8 VM.

```
==> default: Chef 12.5.1 Omnibus package is already installed.
==> default: Running provisioner: chef_solo...
==> default: Detected Chef (latest) is already installed
Uploading chef encrypted data bag secret key...
Generating chef JSON and uploading...
==> default: Running chef-solo...
==> default: stdin: is not a tty
==> default: [2015-12-11T17:35:47+00:00] INFO: Forking chef instance to converge...
==> default: Starting Chef Client, version 12.5.1
==> default: [2015-12-11T17:35:47+00:00] INFO: *** Chef 12.5.1 ***
==> default: [2015-12-11T17:35:47+00:00] INFO: Chef-client pid: 2096
```

**Comment by pietervogelaar** [comment]
I tested with Kitchen, Chef Zero and chef client 12.5.1. But that worked fine. Can you test if it works for you by changing line 17  "template_cookbook = cookbook_name" to "template_cookbook = cookbook_name.to_s"?

**Comment by nickbroon** [comment]
That seems to have fixed that problem, thanks. It now get further.

```
==> default: Chef::Exceptions::PrivateKeyMissing
==> default:
==> default: -----------------------------------
==> default:
==> default: I cannot read /etc/chef/client.pem, which you told me to use to sign requests!
==> default:
==> default:
==> default:
==> default:
==> default: Cookbook Trace:
==> default:
==> default: ---------------
==> default:     /tmp/vagrant-chef/7814f25c269eeb8ef5842808bce63998/cookbooks/jenkins-server/recipes/plugins.rb:58:in `from_file'
==> default:     /tmp/vagrant-chef/7814f25c269eeb8ef5842808bce63998/cookbooks/jenkins-server/recipes/default.rb:19:in `from_file'
==> default:     /tmp/vagrant-chef/7814f25c269eeb8ef5842808bce63998/cookbooks/stashykins/recipes/default.rb:10:in `from_file'
==> default:
==> default:   Relevant File Content:
==> default:   ----------------------
==> default:   /tmp/vagrant-chef/7814f25c269eeb8ef5842808bce63998/cookbooks/jenkins-server/recipes/plugins.rb:
==> default:
==> default:    51:
==> default:    52:    service 'jenkins restart' do
==> default:    53:      service_name 'jenkins'
==> default:    54:      action :restart
==> default:    55:    end
==> default:    56:
==> default:    57:    node.set['jenkins_restarted_once'] = true
==> default:    58>>   node.save
==> default:    59:  end
==> default:    60:
==> default:    61:  # Configure plugins
==> default:    62:  Chef::Log.debug '[JENKINS-SERVER] Configure Jenkins plugins'
==> default:    63:
==> default:    64:  node['jenkins-server']['plugins'].each do |plugin, options|
==> default:    65:    if !options['configure'].nil? && (options['configure'] == true || options['configure'] == 'recipe')
==> default:    66:      cookbook = options.key?('cookbook') && !options['cookbook'].nil? ? options['cookbook'] : cookbook_name
==> default:    67:      recipe = options.key?('recipe') && !options['recipe'].nil? ? options['recipe'] : "plugin_#{plugin}"
==> default:
==> default:
==> default:   Running handlers:
==> default: [2015-12-13T21:53:34+00:00] ERROR: Running exception handlers
==> default:   Running handlers complete
==> default: [2015-12-13T21:53:34+00:00] ERROR: Exception handlers complete
==> default:   Chef Client failed. 4 resources updated in 17 seconds
==> default: [2015-12-13T21:53:34+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
==> default: [2015-12-13T21:53:34+00:00] ERROR: I cannot read /etc/chef/client.pem, which you told me to use to sign requests!
==> default: [2015-12-13T21:53:34+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

which looks like I need to either set `default['jenkins-server']['dev_mode']['security']['private_key']` or `default['jenkins-server']['security']['chef-vault']['data_bag_item']`

**Comment by pietervogelaar** [comment]
I think this is also a Chef Solo problem. The cookbook works fine with Chef Zero (the successor of Chef Solo) and Chef Client, but I never took the time to made the cookbook compatible with Chef Solo. Maybe I will in the near feature.

Is it for you possible to use Chef Zero?

**Comment by nickbroon** [comment]
I ported my Vagrantfile to use chef_zero, and don't get many of the issue.
It would great to document which of the node attributes are strictly required.

**Comment by pietervogelaar** [comment]
Okay, great!

**Comment by pkyeck** [comment]
running into the same problem here - also still using chef solo. tried to switch to chef zero but having problems. do you maybe have some links where I can find some more info on how to make the switch / use chef zero?

btw: chef zero + vagrant is not the problem ... but how can I "cook remotely"?

**Comment by vogentis-vagrant** [comment]
I just released version 0.6.0 which has Chef Solo support!

**Comment by pietervogelaar** [comment]
I just released version 0.6.0 which has Chef Solo support!
